servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImpl.java

```diff
@@ -308,11 +308,17 @@ public Reference assignReference(
       ReferenceType referenceType, String referenceName, String expectedHash, Reference assignTo)
       throws NessieNotFoundException, NessieConflictException {
     try {
+      ReferenceInfo<CommitMeta> resolved =
+          getStore().getNamedRef(referenceName, GetNamedRefsParams.DEFAULT);
 
       ResolvedHash oldRef =
           getHashResolver()
               .resolveHashOnRef(
-                  referenceName, expectedHash, "Expected hash", REQUIRED_UNAMBIGUOUS_HASH);
+                  resolved.getNamedRef(),
```

Comment:
Isn't this change functionally equivalent to the old version? Except for the detached use case... I suspect you are trying to special-case detached refs here, but why?

(BTW this change is probably why we are getting the strange 404 error in one of the test cases above.)

Reply:
The problem is that assigning without an "expected hash" used to work, which is incorrect IMHO. This is case I was trying to fix. I'll see if I can find a better solution.

Reply:
I've added a simple `checkArgument` for that use case for now on top of old reference resolution logic.